### PR TITLE
Adding 'bitorder' support to `cupy.unpackbits` 

### PR DESCRIPTION
**Pull request by iskode**
This PR is about this issue: https://github.com/cupy/cupy/issues/5762

### CODE REVIEW

File: cupy/_binary/packing.py [hunk 1]
```diff
@@ -42,21 +42,28 @@ def packbits(myarray):
...
-def unpackbits(myarray):
+def unpackbits(myarray, bitorder='big'):
     """Unpacks elements of a uint8 array into a binary-valued output array.
 
     This function currently does not support ``axis`` option.
 
     Args:
         myarray (cupy.ndarray): Input array.
 
+        bitorder : {'big', 'little'}, optional. Defaults to 'big'
```
Comment (emcastillo) [inline cupy/_binary/packing.py:65]:
```suggestion
        bitorder (str, optional): bit order to use when unpacking the array, allowed values are `'little'` and `'big'`. Defaults to `'big'`.
```

Comment (emcastillo) [reply cupy/_binary/packing.py:65]: minor fix to docs to keep format consistent.

File: tests/cupy_tests/binary_tests/test_packing.py [hunk 2]
```diff
@@ -39,3 +40,17 @@ def test_unpackbits(self):
...
+
+    def test_pack_unpack_order(self):
+        for bo in ['big', 'little']:
+            self.check_unpackbits([], bitorder=bo)
+            self.check_unpackbits([0], bitorder=bo)
+            self.check_unpackbits([1], bitorder=bo)
+            self.check_unpackbits([255], bitorder=bo)
+            self.check_unpackbits([100, 200, 123, 213], bitorder=bo)
+        a = cupy.array([10, 20, 30], dtype=cupy.uint8)
+        pytest.raises(ValueError, cupy.unpackbits, a, bitorder='r')
```
Comment (emcastillo) [inline tests/cupy_tests/binary_tests/test_packing.py:55]: can you separate these two assers into a new test?

`def test_unpack_invalid_order` or something similar :).
Thanks!

File: tests/cupy_tests/binary_tests/test_packing.py [hunk 3]
```diff
@@ -39,3 +40,17 @@ def test_unpackbits(self):
...
+        pytest.raises(TypeError, cupy.unpackbits, a)
+        pytest.raises(TypeError, cupy.unpackbits, a.astype(float))
+
+    def test_pack_unpack_order(self):
+        for bo in ['big', 'little']:
+            self.check_unpackbits([], bitorder=bo)
+            self.check_unpackbits([0], bitorder=bo)
+            self.check_unpackbits([1], bitorder=bo)
+            self.check_unpackbits([255], bitorder=bo)
+            self.check_unpackbits([100, 200, 123, 213], bitorder=bo)
```
Comment (emcastillo) [inline tests/cupy_tests/binary_tests/test_packing.py:53]: seems like check is not asserting that the bits where correctly unpacked. Can you add a check that verifies that the unpacking was effectively done in big or little endian?

**Review by emcastillo — COMMENTED** [review]
In general looks great! I only have some doubts about the tests :)
Thanks again

File: tests/cupy_tests/binary_tests/test_packing.py (hunk 2 above)
Comment (iskode) [reply tests/cupy_tests/binary_tests/test_packing.py:55]: Thank you for the suggestion. I did the same for the two asserts I've added to `test_unpackbits` in order to assert the array validity.

File: tests/cupy_tests/binary_tests/test_packing.py (hunk 3 above)
Comment (iskode) [reply tests/cupy_tests/binary_tests/test_packing.py:55]: Here we're checking against numpy counterpart implementation following `test_unpackbits`
pattern. So if 'big' endian from cupy matches that of numpy then we're doing a good job.
Should I not test against numpy in the case of bitorder ?

Comment (emcastillo) [reply tests/cupy_tests/binary_tests/test_packing.py:55]: ah sorry, i did not properly read the check code, then this is fine!!

File: cupy/_binary/packing.py [hunk 4]
```diff
@@ -42,21 +42,28 @@ def packbits(myarray):
...
-def unpackbits(myarray):
+def unpackbits(myarray, bitorder='big'):
     """Unpacks elements of a uint8 array into a binary-valued output array.
 
     This function currently does not support ``axis`` option.
 
     Args:
         myarray (cupy.ndarray): Input array.
 
+        bitorder (str, optional): bit order to use when unpacking the array, allowed values are `'little'` and `'big'`. Defaults to `'big'`.
```
Comment (emcastillo) [inline cupy/_binary/packing.py:65]:
```suggestion
        bitorder (str, optional): bit order to use when unpacking the array,
        allowed values are `'little'` and `'big'`. Defaults to `'big'`.
```

Comment (emcastillo) [reply cupy/_binary/packing.py:65]: my bad :D

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by iskode** [comment]
Hi @emcastillo ,
sorry :( about my answer concerning the clarity of 'big'/'little' endian tests. Now thinking back, I believe this test pattern is not good. We must have tests with clear expected values asserted not only against `numpy`. That means if there is a bug on numpy we should reproduce it and also it is a burden to drag this dependency as the library evolves.
Thus I'd like to refactor the whole file `test_packing` by adding tests free from numpy.
What is your take on this ?

**Comment by emcastillo** [comment]
Hey, sorry for my late reply!!!

> We must have tests with clear expected values asserted not only against numpy. That means if there is a bug on numpy we should reproduce it and also it is a burden to drag this dependency as the library evolves.

Yeah, this totally makes sense, but we don't need to go that far, actually, we try to guarantee numpy compatibility so just testing against numpy is fine, we do it with all routines. Thanks!!

**Pull request merged**